pyro/poutine/reparam_messenger.py

```diff
@@ -44,20 +54,64 @@ def _pyro_sample(self, msg):
         if reparam is None:
             return
 
+        # See https://github.com/pyro-ppl/pyro/issues/2878
+        # This is a tricky hack to apply messengers in an order other than the
+        # standard _PYRO_STACK order. Our goal is to allow (model, initializer)
+        # pairs to be reparametrized as a unit. The problem is that messengers
+        # are typically applied in the order
+        #
+        #    InitMessenger(init_to_value(...))(ReparamMessenger(...)(model))
+        #
+        # so that original model sites are reparametrized by the time they are
+        # seen by init_to_value(). To work around this we allow
+        # ReparamMessenger to apply enclosing InitMessengers early, simulating
+        # a priority system for messengers (indeed we might consider
+        # prioritizing messengers). Note that the enclosing InitMessenger will be
+        # called a second time, after ReparamMessenger, but that is ok because
+        # InitMessenger does not overwrite values.
+        #
+        # To get this same logic to work for ConditionMessenger or
+        # ReplayMessenger we would need to ensure those messengers can
+        # similarly be safely applied twice, with the second application
+        # avoiding overwriting the original application.
+        for m in _get_init_messengers():
+            m._pyro_sample(msg)
+
+        # Pass args_kwargs to the reparam via a side channel.
         reparam.args_kwargs = self._args_kwargs
         try:
-            new_fn, value = reparam(msg["name"], msg["fn"], msg["value"])
+            new_msg = reparam.apply({
+                "name": msg["name"],
+                "fn": msg["fn"],
+                "value": msg["value"],
+                "is_observed": msg["is_observed"],
+            })
         finally:
             reparam.args_kwargs = None
 
-        if value is not None:
-            if msg["value"] is None:
-                msg["is_observed"] = True
-            msg["value"] = value
+        if new_msg["value"] is not None:
+            # Validate while the original msg["fn"] is known.
             if getattr(msg["fn"], "_validation_enabled", False):
-                # Validate while the original msg["fn"] is known.
-                msg["fn"]._validate_sample(value)
-        msg["fn"] = new_fn
+                msg["fn"]._validate_sample(new_msg["value"])
+
+            if msg["value"] is not None and msg["value"] is not new_msg["value"]:
+                # Check that overwritten initialization preserves shape.
+                if not torch._C._get_tracing_state():
+                    assert new_msg["value"].shape == msg["value"].shape
+
+                # Warn if a custom init method is overwritten by another init method.
```

Comment:
Sounds reasonable to me.

Reply:
Yeah, this is from personal experience where I spent about 6 hours hunting down a bug that ended up being overwritten initialization 😅